File: DailyLeetcode/April/April19.cpp
#include <vector>
using namespace std;
/*
 * Classic DFS question, but this is a bit dumb -> better one is the rust
 * version
 */
class Solution {
public:
  int numIslands(vector<vector<char>> &grid) {
    int result = 0;
    for (int i = 0; i < grid.size(); i++) {
      for (int j = 0; j < grid[0].size(); j++) {
        if (dfs(grid, i, j) > 0) {
          result++;
        }
      }
    }
    return result;
  }

  int dfs(vector<vector<char>> &grid, int i, int j) {
    int count = 0;
    if (i < 0 || i >= grid.size() || j < 0 || j >= grid[0].size() ||
        grid[i][j] != '1') {
      return 0;
    }

    // Mark the current cell as visited
    grid[i][j] = '#';

    count++;
    count += dfs(grid, i - 1, j);
    count += dfs(grid, i + 1, j);
    count += dfs(grid, i, j - 1);
    count += dfs(grid, i, j + 1);
    return count;
  }
};

File: src/DailyLeetcode/April/April19.rs
pub struct Solution {}

impl Solution {
    pub fn num_islands(mut grid: Vec<Vec<char>>) -> i32 {
        let mut result = 0;
        for i in 0..grid.len() {
            for j in 0..grid[0].len() {
                if grid[i][j] == '1' {
                    result += 1;
                    Self::dfs(&mut grid, i, j);
                }
            }
        }
        result
    }

    fn dfs(grid: &mut Vec<Vec<char>>, i: usize, j: usize) {
        if i < 0 || i >= grid.len() || j < 0 || j >= grid[0].len() || grid[i][j] != '1' {
            return;
        }

        grid[i][j] = '#';
        Self::dfs(grid, i + 1, j);
        Self::dfs(grid, i - 1, j);
        Self::dfs(grid, i, j - 1);
        Self::dfs(grid, i, j + 1);
    }
}
